airbyte-workers/src/main/java/io/airbyte/workers/WorkerApp.java

```diff
@@ -0,0 +1,146 @@
+import org.slf4j.MDC;
+
+public class WorkerApp {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(WorkerApp.class);
+  public static final int KUBE_HEARTBEAT_PORT = 9000;
+
+  private final Path workspaceRoot;
+  private final ProcessFactory processFactory;
+  private final WorkflowServiceStubs temporalService;
+  private final MaxWorkersConfig maxWorkers;
+
+  public WorkerApp(Path workspaceRoot,
+                   ProcessFactory processFactory,
+                   WorkflowServiceStubs temporalService,
+                   MaxWorkersConfig maxWorkers) {
+    this.workspaceRoot = workspaceRoot;
+    this.processFactory = processFactory;
+    this.temporalService = temporalService;
+    this.maxWorkers = maxWorkers;
+  }
+
+  public void start() {
+    Map<String, String> mdc = MDC.getCopyOfContextMap();
+    Executors.newSingleThreadExecutor().submit(
+        () -> {
+          MDC.setContextMap(mdc);
+          try {
+            new WorkerHeartbeatServer(KUBE_HEARTBEAT_PORT).start();
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        });
+
+    final TemporalPool temporalPool = new TemporalPool(temporalService, workspaceRoot, processFactory, maxWorkers);
+    temporalPool.run();
+  }
+
+  private static ProcessFactory getProcessBuilderFactory(Configs configs) throws IOException {
+    if (configs.getWorkerEnvironment() == Configs.WorkerEnvironment.KUBERNETES) {
+      final ApiClient officialClient = Config.defaultClient();
+      final KubernetesClient fabricClient = new DefaultKubernetesClient();
+      final String localIp = InetAddress.getLocalHost().getHostAddress();
+      final String kubeHeartbeatUrl = localIp + ":" + KUBE_HEARTBEAT_PORT;
+      LOGGER.info("Using Kubernetes namespace: {}", configs.getKubeNamespace());
+      return new KubeProcessFactory(configs.getKubeNamespace(), officialClient, fabricClient, kubeHeartbeatUrl);
+    } else {
+      return new DockerProcessFactory(
+          configs.getWorkspaceRoot(),
+          configs.getWorkspaceDockerMount(),
+          configs.getLocalDockerMount(),
+          configs.getDockerNetwork());
+    }
+  }
+
+  public static void waitForServer(Configs configs) throws InterruptedException {
+    final AirbyteApiClient apiClient = new AirbyteApiClient(
+        new io.airbyte.api.client.invoker.ApiClient().setScheme("http")
+            .setHost(configs.getAirbyteApiHost())
+            .setPort(configs.getAirbyteApiPort())
+            .setBasePath("/api"));
+
+    boolean isHealthy = false;
+    while (!isHealthy) {
+      try {
+        HealthCheckRead healthCheck = apiClient.getHealthApi().getHealthCheck();
+        isHealthy = healthCheck.getDb();
+      } catch (ApiException e) {
+        LOGGER.info("Waiting for server to become available...");
+        Thread.sleep(2000);
+      }
+    }
+  }
+
+  public static void main(String[] args) throws IOException, InterruptedException {
```

Comment:
What about merging the TemporalPool and WorkerApp classes together? There doesn't seem to be much benefit keeping them as separate concepts in the code.

Reply:
I'd lean towards getting rid of TemporalPool (the name is confusing) and bringing everything into the worker app.

Feels more natural for all concepts to be in workers terms.

Reply:
agreed